feat: Table widget checkbox column type

keyurparalkar opened the pull request:
## Description

This PR introduces a new column type: `Checkbox`. It uses checkbox widget's component to render the checkbox inside the table's cell. 

Here is the list of changes that are implemented:
* New column type added: Checkbox.
* Property pane level options:
     * Computed Value
     * Visible
     * Editable
     * Styles:
          * Horizontal Alignment
          * Vertical Alignment
          * Cell background
     * Events:
          * onCheckChange: This is available only when editable is true
* Table can now be filtered with is Checked and is unChecked conditions

**_NOTE_**: When the Editable property is false then the checkboxes are disabled.

Please refer to the following for more information:
* [PRD](https://www.notion.so/appsmith/Checkbox-in-table-687ad8bcf19243428c17ad5f7af9e0c4)
* [Figma](https://www.figma.com/file/aJ2Hwz9gH5AK6zaRYtZV8a/Table-widget-update?node-id=527%3A46002)

Fixes #7338 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

* Cypress spec: `checkboxCell_spec.js` is added. It covers the following scenarios:
    * Checks if the checkbox column type is available.
    * Checks if the computed value is editable and `currentRow` object is available.
    * Verifies visibility of the column
    * When editable is `true` then check:
         * If check box is enabled 
         * If `onCheckChange` is available
         * If handler attached to `onCheckChange` is working via interacting with the checkbox
    * if editable is `false` then verify if checkboxes are disabled
    * Check horizontal, vertical alignment and cell background color
    * Check if the filter conditions: `is checked` and `is unchecked` are working as expected.
    * Verify that `onCheckChange` is not available on `custom columns`

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,133 @@
+  );
+};
+
+export const CheckboxCellWrapper = (props: renderCellType) => {
```
CheckboxCell

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,133 @@
+    >
+      {columnActions.map((action: ColumnAction, index: number) => {
+        return (
+          <Checkbox
```
Directly use the checkbox widget component

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/index.tsx:
```diff
@@ -1493,6 +1501,45 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
+                rowIndex,
+                action,
+                onComplete,
+                triggerPropertyName: "onCheckChange",
```
onCheckedChange

keyurparalkar replied:
Since checkbox widget is also using `onCheckChange` therefore using the same terminology over here as well

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/index.tsx:
```diff
@@ -1536,6 +1583,13 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
     }
   };
 
+  onCheckboxChange = (alias: string, value: boolean, rowIndex: number) => {
```
Directly call the updateTransientTableData function

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/ColumnControl.ts:
```diff
@@ -64,6 +64,10 @@ export default {
         },
+        {
+          label: "Checkbox",
+          value: "checkbox",
```
use columnType enums

keyurparalkar commented:
@sbalaji1192 Please refer to the below information about `cellComponentHorizontalAlignment`:

* This is a new column level property that is added. This prop will be used to manage horizontal align for the checkbox component or switch component. 
* Working
    * We have two types of horizontal alignments now:
        * Column level(`cellComponentHorizontalAlignment` prop)
        * Table level(`horizontalAlignment` prop)
    * Scenario 1: When updating horizontal alignment at table level:
        * While updating at table level we will also need to update the `cellComponentHorizontalAlignment` in sync with `horizontalAlignment`.
        * To do this, I have added a hook named: `updateCellComponentHorizontalAlignment` to the `Text Align` prop at the table level property pane config.
        * This hook will make sure that whenever we update the `horizontalAlignment` it will remove the `cellComponentHorizontalAlignment` from the column. In this way, the `HeaderCell` and `CheckboxCell` will make use of the `horizontalAlignment` when changed at table level.
     * Scenario 2: When updating horizontal alignment at table level:
        * Whenever it is changed at column level, then it will make use of `cellComponentHorizontalAlignment`.
     * These are the lines which helps to decide what alignment to pickup in the above scenarios: 
            * [HeaderCell](https://github.com/appsmithorg/appsmith/blob/c3d9febf4e7c9f79965547e3e2326e7ecb150247/app/client/src/widgets/TableWidgetV2/component/cellComponents/HeaderCell.tsx#L134) 
            * [CheckboxCell](https://github.com/appsmithorg/appsmith/blob/c3d9febf4e7c9f79965547e3e2326e7ecb150247/app/client/src/widgets/TableWidgetV2/widget/index.tsx#L1513)

* The `updateCellComponentHorizontalAlignment` also sets the `cellComponentHorizontalAlignment` to `CENTER` whenever a new column is added to the table.

sbalaji1192 commented on app/client/package.json:
```diff
@@ -237,7 +237,7 @@
     "@typescript-eslint/parser": "^5.25.0",
     "babel-plugin-module-resolver": "^4.1.0",
-    "babel-plugin-styled-components": "^2.0.7",
+    "babel-plugin-styled-components": "2.0.7",
```
is this intended?

keyurparalkar replied:
Changed it to `"^2.0.7"`

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,98 @@
+export const CheckboxCell = (props: CheckboxCellProps) => {
+  const {
+    accentColor,
+    allowCellWrapping,
```
We don't need allowCellWrapping

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,98 @@
+    borderRadius,
+    cellBackground,
+    cellComponentHorizontalAlignment,
+    columnAction,
```
don't need column action

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,98 @@
+    isHidden,
+    onCommandClick,
+    textColor,
+    textSize,
```
we don't need these two textColor & textSize

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,98 @@
+        isDisabled={isDisabled}
+        isLoading={false}
+        isRequired={false}
+        key={columnAction.id}
```
key is not needed

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,98 @@
+        label=""
+        labelPosition={LabelPosition.Auto}
+        onCheckChange={handleChange}
+        rowSpace={5}
```
remove

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/index.tsx:
```diff
@@ -1503,6 +1508,49 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
+                onComplete,
+                triggerPropertyName: "onCheckChange",
+                eventType: EventType.ON_CHECK_CHANGE,
+              });
```
const row = filteredTableData[rowIndex];
{
...row,
[alias]: !props.cell.value,
}

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/propertyConfig/Styles.ts:
```diff
@@ -92,7 +96,10 @@ export default {
       label: "Text Align",
       controlType: "ICON_TABS",
-      updateHook: updateColumnStyles,
+      updateHook: composePropertyUpdateHook([
```
instead of updateCellComponentHorizontalAlignment, lets extend label property to accept dynamic strings.

sbalaji1192 commented on app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx:
```diff
@@ -425,7 +425,10 @@ const PropertyControl = memo((props: Props) => {
-  const { label, propertyName } = props;
+  const { propertyName } = props;
+  const label = isFunction(props.label)
+    ? props.label(widgetProperties, props.propertyName)
```
use `propertyName`

sbalaji1192 commented on app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx:
```diff
@@ -425,7 +425,10 @@ const PropertyControl = memo((props: Props) => {
+  const { propertyName } = props;
+  const label = isFunction(props.label)
+    ? props.label(widgetProperties, props.propertyName)
+    : props.label;
```
leave a line space here

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,81 @@
+  value: boolean;
+  accentColor: string;
+  isDisabled?: boolean;
+  onCommandClick: () => void;
```
lets call this onChange

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/Checkbox.tsx:
```diff
@@ -0,0 +1,81 @@
+import React from "react";
```
lets call this file checkbox cell and other one selectionCheckboxCell

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/index.tsx:
```diff
@@ -1503,6 +1508,46 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
+            isHidden={isHidden}
+            onCommandClick={() => {
+              const originalIndex = this.getRowOriginalIndex(rowIndex);
+              const row = filteredTableData[originalIndex];
```
use rowIndex here

chandannkumar commented:
Observations @keyurparalkar 
1. All the checkbox are checked by default on Deploy preview (view mode) irrespective of configurations done on edit mode
2. Filter for filtering data based on isChecked/isUnchecked conditions is not giving correct results
3. 'Computed Value' property shows previous column values until updated once
4. No autocomplete suggestion for {{currentRow}} on checkbox type column

 Video for all : https://www.loom.com/share/3f1910def71646b0b3a3feb54e353a6b


Queries: @dilippitchika 
1. Are we not providing 'Required' property as mentioned on PRD?
2. Can we add '**Default Selected**' property to handle default checked issue(point 1 on Observations) ?

keyurparalkar commented:
@chandannkumar Please find below the comments for the above QA callouts:

> All the checkbox are checked by default on Deploy preview (view mode) irrespective of configurations done on edit mode

* There are two cases to this:
    * Converting a non-boolean value to checkbox type:
        * In this case, since its a non-boolean value like string then all the checkboxes will be checked.
    * Converting boolean value to checkbox type:
         * Then it will respect the boolean value.
     * For both of the above cases, if you try to enable editing of a column, do any edits and then deploy the app, then this edited data won’t persist. Since we moved to another page that is the deployed app therefore, the edited data is lost.
     * You can observe a similar behaviour on Plain Text columns

> Filter for filtering data based on isChecked/isUnchecked conditions is not giving correct results

* This is an expected. Values to be filtered should be of boolean type and not non-boolean values.
* 

> 'Computed Value' property shows previous column values until updated once

* This is an expected behaviour

> No autocomplete suggestion for {{currentRow}} on checkbox type column

* Unable to reproduce this issue.

@chandannkumar I have fixed the disabled cursor flickering issue as well.

dilippitchika commented:
1. Are we not providing 'Required' property as mentioned on PRD?
Required for this doesn't make sense yet, we are waiting for some user feedback to figure out if we need to add this
2. Can we add 'Default Selected' property to handle default checked issue(point 1 on Observations) ?
Here the computed value is the default selected property, so based on table data the default selection will be on or off

@chandannkumar

chandannkumar commented:
@keyurparalkar 
Checkbox are unclickable when column is converted to boolean from non-boolean value
https://www.loom.com/share/35d9b897524c4315aeeea9aee6bff0ae

keyurparalkar commented:
> 

@chandannkumar This is an expected behaviour.

chandannkumar commented:
- On changing checkbox to true/false, it seems that Save/Discard column is getting activated but without Save or Discard it is allowing to edit next cell as well.
Since in table v2 for other column type like plain text, editing is restricted for next cell until current cell changes are Saved or Discarded

https://www.loom.com/share/44f9a63be721497b8e3654e1c5071883

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/component/cellComponents/CheckboxCell.tsx:
```diff
@@ -1,53 +1,132 @@
+          hoverOpenDelay={200}
+          position="top"
+        >
+          <CheckboxComponent
```
Create a const with CheckboxComponent and use it in two places

keyurparalkar commented:
> * On changing checkbox to true/false, it seems that Save/Discard column is getting activated but without Save or Discard it is allowing to edit next cell as well.
>   Since in table v2 for other column type like plain text, editing is restricted for next cell until current cell changes are Saved or Discarded
> 
> https://www.loom.com/share/44f9a63be721497b8e3654e1c5071883

@chandannkumar This is fixed now. Please re-test

keyurparalkar commented:
/ok-to-test sha=146a3c3

laveena-en commented:
Tested and verified.

keyurparalkar commented:
/ok-to-test sha=146a3c3

sbalaji1192 commented:
/ok-to-test sha=146a3c3

Aishwarya-U-R commented:
/ok-to-test sha=db1e670

keyurparalkar commented:
/ok-to-test sha=53152b6

chandannkumar commented:
Tested this PR and working as expected

keyurparalkar commented:
/ok-to-test sha=3ef83f0

keyurparalkar commented:
/ok-to-test sha=a0bd009

sbalaji1192 commented on app/client/src/pages/Editor/PropertyPane/Generator.tsx:
```diff
@@ -61,7 +61,7 @@ function Section(props: SectionProps) {
         collapsible={sectionConfig.collapsible ?? true}
         hidden={sectionConfig.hidden}
-        id={config.id || sectionConfig.sectionName}
+        id={config.id || (sectionConfig.sectionName as string)}
```
why do we need to use as keyword?

sbalaji1192 commented on app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/index.ts:
```diff
@@ -526,7 +544,13 @@ export default {
+          label: (props: TableWidgetProps, propertyPath: string) => {
+            const basePropertyPath = getBasePropertyPath(propertyPath);
+            const columnType = get(props, `${basePropertyPath}.columnType`);
+            return columnType === "checkbox"
```
use Columntypes enum here instead of literal "checkbox"

keyurparalkar commented:
/ok-to-test sha=551e989

Aishwarya-U-R commented:
/ok-to-test sha=e5014c3

keyurparalkar commented:
/ok-to-test sha=434aa93
